Actions System + UI

chairbender opened the pull request:
Fixes #1945 

Adds a system to define, grant, revoke, and control actions. Adds the associated hotbar and action picker window, accessible by clicking the settings icon. Implements scream, toggle internals, toggle light actions along with various debug actions.
Also makes some improvements to alerts system based on stuff I figured out for actions system.

To see the debug item actions, you can spawn and pickup a "item action example" entity. You can manually grant / revoke player (non-item) debug actions using the new commands grantaction, revokeaction, and coolaction (to set cooldowns). For example, `grantaction DebugInstant`.

At the moment, though I have tested it manually, there is no automatic test coverage whatsoever. I wanted to get the review started first to improve the chances of this making the playtest, so I can add test coverage as I address feedback (possibly in a future PR so it doesn't hold this one up). One way or another I will ensure there's test coverage.

I would suggest you start reviewing from the new BaseActionPrototype, ItemActionPrototype, ActionPrototype. The definitions for the prototypes are in actions.yml and item_actions.yml. Then head over to SharedActionsComponent to see the main logic of how action states are managed. The ClientActionsComponent is the main entry point for all the clientside actions UI, so you should start there to see how the UI code is working.

Action functionality:
- Actions can be implemented via one of 4 "behaviors" - instant, target (point/entity), toggle, each with associated interfaces
- Target-based actions can be set "repeatable" so the user selects the action and can then click to use it multiple times until deselecting it.
- Item-based actions can be displayed in a few ways (wasn't sure what was best), controlled via iconStyle. Right now I like BigItem best so it's the default. You can see all of them in action on "item action example" entity.
- Actions have an associated name, description, requirements text, which can accept formatting.
- Toggle actions can have a toggle on / off icon or just highlight.
- filters and search keywords can be defined per action to improve searchability
- Action assignments can be dragged and dropped in the hotbar / from the action window
- Mousing over item actions highlights the corresponding slot
- Hotbar uses a 2 column layout if screen is too short
- Lock / unlock to prevent clearing / dragging actions
- When an action is newly-granted, it will auto-populate the hotbar (overflowing to next hotbar if full)
- Actions can be assigned to multiple slots
- When a user clears an action by right clicking, it will no longer auto-populate, they must use the action window to reassign it. As a special case, I wanted to ensure that users are aware of new actions when they pick up an item, so item actions which are cleared this way are only cleared while the item is still in inventory
- Theoretically I tried to let it support prediction - all of the commands to modify the action statuses can be done on the shared component, but in practice I found that the client state is always immediately overridden by an older server state by ResetPredictedEntities. Also a lot of the associated interaction interfaces / messages (equip / unequip) only fire server side currently.
- Actions can be disabled/revoked temporarily, so theoretically you could use the action blocker system to say, disable the action while player is stunned. But this is complicated and action-specific, so right now we do not use this (the blocker system can still be checked in the action implementation to prevent actually doing the action behavior, but the action will appear usable by the player even though it's technically not).
- Item actions that have a cooldown are currently only tied to the combo of player+item (see the debug instant action on the "item action example", try dropping and picking it up). Another player picking up the item would not have the cooldown.

Other stuff:
- Added interfaces for hand equip / unequip and clarified comments about when all the equipment stuff is called. I will be updating the wiki with this info as well.
- refactored drag drop logic into DragDropHelper to avoid code duplication.
- Adds tgstation screams and a few icons for the new actions

Future things:
- configurable auto-population behavior
- Hotbar presets / persistence between rounds (so if you like things being set up a certain way you could have that setup persist, saved somewhere locally, possibly per-role)
- optional autocast for targeted interactions - instead of clicking the action then a target, you would just press the action hotkey and it would immediately target wherever your mouse is currently over (like shift+1 by default to autocast slot 1).

Video
https://cdn.discordapp.com/attachments/310560938438819861/784927196309094450/DSaDDNIpfW.mp4
https://cdn.discordapp.com/attachments/310560938438819861/784927221890285588/7RfYpgN2vn.mp4


Depends on space-wizards/RobustToolbox#1435

metalgearsloth commented:
Might need an update with master (same with RobustToolbox) I think as when I try to checkout this + the RT commit it fails.

metalgearsloth reviewed (commented):
Some general thoughts from using it ingame (mostly nice-to-haves so not urgent to do), bearing in mind these are all my opinions:
1. Should be able to drag the hotbar icons off to remove them (this is how it normally works in MMOs).
2. Highlights on all hovers IMO.
3. Make the left and right arrows to swap hotbars have wider hitboxes.
4. So currently in the bottom-right if the action isn't available it shows the icon which is nice. IMO that should go away when the slot is filled (it just looks kinda weird ingame).
edit: Okay I'm special that's because it's the action icon right? Maybe hide it if it has the same path as the item icon? eh not sure on that so maybe ignore it.
5. Maybe add a help button to the book thing just with some tips or whatever (would be more visible than putting it on the F1 menu IMO).
6. "All" being a filter is kind of weird given I'd expect it to enable everything when selected but it seems to do its own thing?
7. Probably related to 2 but dragging onto the hotbar should hover if it's valid to drop there.
8. Were the keybinds added to the controls screen? Couldn't seem to see them.

When I was clicking on the flashlight in the hotbar it didn't work but the one in the inventory worked and lit it up which was weird.

This is all really nitpicky stuff but it's still really good (will look at the code soon tm).

DrSmugleaf commented on Content.Client/GameObjects/Components/Mobs/Actions/ActionAssignments.cs:
```diff
@@ -0,0 +1,300 @@
+        /// auto-populate.
+        /// </summary>
+        private HashSet<ActionType> _preventAutoPopulate = new HashSet<ActionType>();
+        private Dictionary<EntityUid,HashSet<ItemActionType>> _preventAutoPopulateItem =
```
```suggestion
        private Dictionary<EntityUid, HashSet<ItemActionType>> _preventAutoPopulateItem =
```

DrSmugleaf commented on Content.Client/GameObjects/Components/Mobs/Actions/ActionAssignment.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Content.Client.GameObjects.Components.Mobs.Actions
+{
+    public struct ActionAssignment
```
Could implement IEquatable\<ActionAssignment\> here.

DrSmugleaf commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,981 @@
+            var uiManager = IoCManager.Resolve<IUserInterfaceManager>();
+            uiManager.StateRoot.AddChild(_ui);
+
+            _inventoryComponent = Owner.GetComponent<ClientInventoryComponent>();
+            _handsComponent = Owner.GetComponent<HandsComponent>();
```
This will error if these two components don't exist. Replace it with a TryGet, or better yet change `_inventoryComponent` and `_handsComponent` to use component dependencies.

metalgearsloth commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,911 @@
+        {
+            base.HandleComponentState(curState, nextState);
+
+            if (!(curState is ActionComponentState state))
+            {
```
```suggestion
            if (curState is not ActionComponentState state)
            {
```

metalgearsloth commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,911 @@
+            var toRemove = new List<(EntityUid item, ItemActionType actionType)>();
+            foreach (var itemActionCooldown in _itemActionCooldowns)
+            {
+                var expiryTime = GameTiming.CurTime - itemActionCooldown.Value.end;
+                if (expiryTime > CooldownExpiryThreshold)
```
```suggestion
                if ((GameTiming.CurTime - itemActionCooldown.Value.end) > CooldownExpiryThreshold)
```

metalgearsloth commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,911 @@
+    [Serializable, NetSerializable]
+    public class PerformToggleActionMessage : PerformActionMessage
+    {
+        /// <summary>
+        /// True if we are trying to toggle the action on, false if trying to toggle it off.
+        /// </summary>
+        public readonly bool ToggleOn;
+
+        public PerformToggleActionMessage(ActionType actionType, bool toggleOn) : base(actionType)
+        {
+            ToggleOn = toggleOn;
+        }
```
So what you can do for these to save bandwidth is have a message for on and a message for off. It makes the code clunkier but you save the extra bool being sent.

metalgearsloth commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,911 @@
+    /// <summary>
+    /// A message that tells server we want to toggle the indicated action.
+    /// </summary>
+    [Serializable, NetSerializable]
+    public class PerformToggleItemActionMessage : PerformItemActionMessage
+    {
+        /// <summary>
+        /// True if we are trying to toggle the action on, false if trying to toggle it off.
+        /// </summary>
+        public readonly bool ToggleOn;
+
+        public PerformToggleItemActionMessage(ItemActionType actionType, EntityUid item, bool toggleOn) : base(actionType, item)
+        {
+            ToggleOn = toggleOn;
+        }
+    }
```
Same here

metalgearsloth reviewed (changes requested):
![image](https://user-images.githubusercontent.com/31366439/101282307-8bc93a00-3828-11eb-838e-fd6ea499f3b3.png)

My IDE's giving me a few warnings around the place on some stuff that seems slightly sketchy so you might want to suss it out. You also use static readonly in a few areas when you could use consts (in one area I left a comment on you also get a bonus out of it).

On the audio stuff is it possible to get like a copyright / attribution for it somewhere?

Also nullable on the files would be nice but I'm not gonna twist ya arm over it.

Making this work with prediction would also be a pain so I can understand if you want to out of scope it

chairbender commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,981 @@
+            }
+        }
+
+        public override void HandleComponentState(ComponentState curState, ComponentState nextState)
+        {
+            base.HandleComponentState(curState, nextState);
+
+            if (!(curState is ActionComponentState))
+            {
+                return;
+            }
+
+            UpdateUI();
+        }
```
it happens in the call to base.HandleComponentState

chairbender commented on Content.Client/GameObjects/EntitySystems/DragDropSystem.cs:
```diff
@@ -150,49 +137,142 @@ private bool OnUseMouseDown(in PointerInputCmdHandler.PointerInputCmdArgs args)
+            {
+                return false;
+            }
+            // still in range of the thing we are dragging?
+            if (!_interactionSystem.InRangeUnobstructed(_dragger, _dragDropHelper.Dragged))
+            {
+                return false;
+            }
+
+            // keep dragged entity under mouse
+            var mousePos = _eyeManager.ScreenToMap(_dragDropHelper.MouseScreenPosition);
+            // TODO: would use MapPosition instead if it had a setter, but it has no setter.
+            // is that intentional, or should we add a setter for Transform.MapPosition?
+            _dragShadow.Transform.WorldPosition = mousePos.Position;
+
+            _targetRecheckTime += frameTime;
+            if (_targetRecheckTime > TargetRecheckInterval)
+            {
+                HighlightTargets();
+                _targetRecheckTime = 0;
+            }
+
+            return true;
+        }
```
I saw that when merging. I thought I got those changes incorporated into DragDropHelper but I'll double check.

metalgearsloth replied:
I'll just update mine when I get a chance as I don't think you really changed the underlying functionality.

chairbender commented on Content.Client/GameObjects/Components/Mobs/Actions/ActionAssignments.cs:
```diff
@@ -0,0 +1,300 @@
+                if (assignmentEntry.Key.Assignment != Assignment.ItemActionWithItem) continue;
+                // we have this assignment currently tied to an item,
+                // check if it no longer has an associated item in our dict of states
+                if (itemActionStates.TryGetValue(assignmentEntry.Key.Item.Value, out var states))
+                {
+                    if (states.ContainsKey(assignmentEntry.Key.ItemActionType.Value))
```
will rework ActionAssignments to fix this, it was always annoying

chairbender commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,981 @@
+                        // TODO: This flickers when toggling on due to ResetPredictedEntities being
+                        // called with an older (toggled off) state from the server.
+                        ToggleAction(action.ActionType, !actionState.ToggledOn);
+                        SendNetworkMessage(new PerformToggleActionMessage(action.ActionType,
+                            !actionState.ToggledOn));
```
I don't know about predicted messages. I don't know how I'm actually supposed to implement prediction here and am not super familiar with approaches for doing it in our codebase. I assumed modifying the client state and calling Dirty would be enough, but what happens is there's a call to resetpredictedentities which resets the client state (and I see a TODO: Handle interpolation in that method).

chairbender replied:
doesn't appear to flicker anymore actually

chairbender commented on Content.Shared/Alert/Click/Unbuckle.cs:
```diff
@@ -0,0 +1 @@
+﻿
```
not sure what you mean, this is used when clicking the unbuckle alert (set as its onClick in actions.yml)

chairbender commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,981 @@
+
+            StopHighlightingItemSlots();
+
+            // figure out if it's in hand or inventory and highlight it
+            foreach (var hand in _handsComponent.Hands)
+            {
+                if (hand.Entity == actionSlot.Item && hand.Button != null)
+                {
+                    _highlightingItemSlots.Add(hand.Button);
+                    hand.Button.Highlight(true);
+                    return;
+                }
+            }
+
+            foreach (var slot in _inventoryComponent.AllSlots)
+            {
+                if (slot.Value == actionSlot.Item)
+                {
+                    if (_inventoryComponent.InterfaceController == null) return;
+                    foreach (var itemSlotButton in
+                        _inventoryComponent.InterfaceController.GetItemSlotButtons(slot.Key))
+                    {
+                        _highlightingItemSlots.Add(itemSlotButton);
+                        itemSlotButton.Highlight(true);
+                    }
+                    return;
+                }
+            }
```
couldn't find that, also this is clientside specific and specifically needs access to UI controls, so I'm not really sure how to make this a helper at the moment that would work across client / server (assuming by AI you mean server side AI-controlled mobs)

chairbender commented:
I think I've addressed all the feedback. Any comments about the actual functionality / UI of the bar I'd like to tackle in the next PR along with your suggestions for how the hotbar should actually work from a UI perspective (such as right click not clearing actions).

chairbender commented on Content.Server/Actions/ScreamAction.cs:
```diff
@@ -0,0 +1,78 @@
+                            AudioHelpers.WithVariation(Variation).WithVolume(Volume));
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
```
This was thoughtless of me. For now I will just pick a random sample from all clips regardless of character settings. I am open to other suggestions ofc.

chairbender replied:
From getting input on discord it actually sounds like the current approach is acceptable.

PJB3005 commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,906 @@
+            return false;
+        }
+
+        /// <see cref="TryGetItemActionStates"/>
```
1. You probably meant `<seealso` instead for references like this.
2. This doesn't actually work since you're not specifying any parameters to resolve the overloads.
![image](https://user-images.githubusercontent.com/8107459/101679083-227f3c00-3a5f-11eb-9935-d4394a83d51c.png)

chairbender replied:
I've fixed this, but apparently it won't let me add the parameters to resolve the overloads in the case of named tuples (affects Cooldown overloads).

PJB3005 replied:
Named tuples shouldn't affect overload resolution so using unnamed ones for this probably works?

PJB3005 commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,906 @@
+        /// TODO: Eventually this should probably be a byte so we it can toggle through multiple states.
+        /// </summary>
+        public bool ToggledOn;
+        public ValueTuple<TimeSpan, TimeSpan>? Cooldown;
```
```suggestion
        public (TimeSpan, TimeSpan)? Cooldown;
```

PJB3005 commented on Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs:
```diff
@@ -0,0 +1,906 @@
+        /// Creates an action state for the indicated type, defaulting to the
+        /// initial state.
+        /// </summary>
+        public ActionState(bool enabled = false, bool toggledOn = false, ValueTuple<TimeSpan, TimeSpan>? cooldown = null)
```
```suggestion
        public ActionState(bool enabled = false, bool toggledOn = false, (TimeSpan, TimeSpan)? cooldown = null)
```

PJB3005 commented on Content.Shared/Utility/Cooldowns.cs:
```diff
@@ -0,0 +1,28 @@
+        /// <see cref="FromNow"/>
+        public static (TimeSpan start, TimeSpan end) SecondsFromNow(double seconds, IGameTiming gameTiming = null)
+        {
+            return FromNow(TimeSpan.FromSeconds(seconds));
```
You're not actually using the `gameTiming` you take in here.

PJB3005 reviewed (commented):
In general, I don't like how much code is duplicated between the item and non-item action types.

chairbender commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,956 @@
+
+        /// <summary>
+        /// Allows calculating if we need to act due to this component being controlled by the current mob
+        /// TODO: should be revisited after space-wizards/RobustToolbox#1255
```
fyi @metalgearsloth  (I believe you suggested this TODO on the previous alerts system PR and I just copied it over to this component as well)

chairbender commented on Content.Client/GameObjects/Components/Mobs/ClientActionsComponent.cs:
```diff
@@ -0,0 +1,956 @@
+            Tooltips.PositionTooltip(_tooltip!);
+            // if we set it visible here the size of the previous tooltip will flicker for a frame,
+            // so instead we wait until FrameUpdate to make it visible
+            _tooltipReady = true;
```
turns out when I fixed the way tooltips are implemented as you suggested (new instance for each tooltip), this issue goes away

chairbender commented on Content.Client/UserInterface/Controls/ActionSlot.cs:
```diff
@@ -0,0 +1,495 @@
+        /// <summary>
+        /// Simulates clicking on this, but being done via a keybind
+        /// </summary>
+        public void HandleKeybind(BoundKeyState keyState)
+        {
+            // simulate a click, using UIClick so it won't be treated as a possible drag / drop attempt
+            var guiArgs = new GUIBoundKeyEventArgs(EngineKeyFunctions.UIClick,
+                keyState, new ScreenCoordinates(GlobalPixelPosition), true,
+                default,
+                default);
+            if (keyState == BoundKeyState.Down)
+            {
+                KeyBindDown(guiArgs);
+            }
+            else
+            {
+                KeyBindUp(guiArgs);
+            }
+
+        }
```
@PJB3005 any suggestions for a name for this event?

chairbender commented on Content.Client/UserInterface/ActionsUI.cs:
```diff
@@ -0,0 +1,353 @@
+            LayoutContainer.SetAnchorAndMarginPreset(this, LayoutContainer.LayoutPreset.TopLeft, margin: 10);
+            LayoutContainer.SetMarginTop(this, 100);
+
+            _clyde = IoCManager.Resolve<IClyde>();
```
Just couldn't figure out how to make that approach work, especially not without a "how much space do I have to work with" capability. I did try though. By "resize events" I assume you mean overriding "Resized()"?

If I'm not allowed to look at window size, that leaves me with control size. Resized() doesn't fire unless the control size is being changed by layoutcontainer. So in order to ensure that happens when the window is resized, I need to set a top and bottom anchor using layoutcontainer. Then what happens is, the control expands as it should as I increase window height. I can look at control.Height to see how much height I have to work with, and appropriately choose a single or 2 column layout.

But then if I shrink the height of the window, the control stays "stuck" at its previous height. The min size is calculated based on the contents of my control, which is currently dominated by the height of the single-column grid container inside it. So the control stops receiving Resized() calls and never knows when to switch back to the 2 column layout. Because layout container sees that the min height already overflows the "anchor-defined" height, so decides not to change the height. And my control can't know that it needs to switch to a 2 column layout (thus reducing the min height) because it doesn't have any sense of "how much space do I have to work with"...it just sees that the control size is currently the height of the single-column grid. It doesn't know "oh looks like we're getting squished".

So I don't see a way to accomplish this with the current UI system.

PJB3005 replied:
You could override `CalculateMinimumSize` to return something like 0 to always allow the control to shrink.

chairbender replied:
lol yep works perfectly. In my defense this seems like just as much of a hack as the other approach. But actually it makes sense, I just set the minimum height to the minimum height of the 2 column layout.

PJB3005 replied:
Anecdotally, this would also be fixed by a WPF-style `Measure()` that takes in the available size. But this is a good solution either way.

chairbender commented on Content.Server/GameObjects/Components/Atmos/GasTankComponent.cs:
```diff
@@ -317,6 +330,37 @@ private void CheckStatus()
+            else
+            {
+                actionsComponent.GrantFromInitialState(ItemActionType.ToggleInternals, Owner, false);
+            }
```
It seemed like an interface was the better way to go, which didn't require that component, so I added IItemActionsEquipped and call it from SharedActionsSystem.

chairbender commented on Content.Server/GameObjects/Components/Atmos/GasTankComponent.cs:
```diff
@@ -348,9 +392,16 @@ protected override void Activate(IEntity user, GasTankComponent component)
+            {
+                if (!args.Item.TryGetComponent<GasTankComponent>(out var gasTankComponent)) return;
+                gasTankComponent.ToggleInternals();
+            }
```
this was a pre-existing flaw in GasTankComponent and HandheldLightComponent. Added a CanUse check for both of them so that it will block regardless of how they try to toggle (via action or via Z or bound UI)

chairbender commented:
@PJB3005 alright, I believe I've addressed everything. I also changed the colors to match injazz's new suggestions and added remappable bindings and tutorial info about the new keys. A few other things to call out n addition to the responses I've left on some of the feedback...

For any feedback which also would apply to my earlier alert system work, I have made the fixes there as well.

Code duplication between item / non-item actions was mostly reduced by implementing a common interface in IActionAttempt. Also, I added additional interfaces on messages to eliminate code dupe involved with them.

I pulled a massive amount of UI logic out of the actions component and into the controls themselves. Controls are a lot more self-sufficient, made possible by just providing them ClientActionsComponent and making some things in ClientActionsComponent public.

ActionSlot extends PanelContainer now rather than BaseButton.

Tooltip code is a lot simpler now as well thanks to new tooltip capabilities I added in robust.

PJB3005 reviewed (commented):
Rest is good, but I'm still not quite warm on the way items do actions. I was thinking of, instead of granting to the user, the item just grants to an `ItemActionsComponent` on its entity. Then THAT component is in charge of (de)granting the actions automatically when the item leaves/enters inventory. This should remove all boiler plate on code like `HandheldLightComponent` except the actual code assigning the current action status to the item.

chairbender commented:
> Rest is good, but I'm still not quite warm on the way items do actions. I was thinking of, instead of granting to the user, the item just grants to an `ItemActionsComponent` on its entity. Then THAT component is in charge of (de)granting the actions automatically when the item leaves/enters inventory. This should remove all boiler plate on code like `HandheldLightComponent` except the actual code assigning the current action status to the item.

I was initially nervous about a component-based approach because you would then have to always add the ItemActions component to any entity that provided them. For example, we have a few HandheldLight entities, which now need to have ItemActions added + the definition of the actions it provides. But I guess that minor issue could be solved by prototype inheritance? 

And now that I'm into implementing this, it actually massively cleans up SharedActionsComponent because all the items states can simply be tracked on the item itself.

Should have something in a few hours.

chairbender commented:
@PJB3005 alright, IMO the way item actions are handled is way better now. I have moved the item action state tracking into ItemActionsComponent, which updates its state into the corresponding SharedActionscomponent of the player holding it. There are some other improvements...

For one, cooldowns for item actions are now tracked on the item, which IMO makes more sense than tracking them on the player.

Additionally, I decided to add a bool return to toggle action interfaces. This bool should be true if the toggle attempt was successful. This helps to ensure that no matter what the implementation does, it will explicitly indicate if the toggle attempt was successful and the state should be changed. It's still allowed for the implementation to explicitly modify the toggle status via ItemActionsComponent/SharedActionsComponent. This helps to avoid some mistakes in the implementation of such actions, for example failing an action blocker check and simply returning, without remembering that the toggle status needs to be set back to its old value (since before this change, the toggle action interface was called AFTER already toggling the value).
